ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/blockheaders/NewBlockHeadersSubscriptionServiceTest.java

```diff
@@ -63,76 +73,116 @@
+        blockResultFactory.transactionHash(
+            blockchainQueriesSpy.blockByHashWithTxHashes(forkBlock2.getHash()).orElse(null));
+
+    verify(subscriptionManager, times(3)).notifySubscribersOnWorkerThread(any(), any(), any());
+    verify(subscriptionManager, times(3))
+        .sendMessage(subscriptionIdCaptor.capture(), responseCaptor.capture());
+    assertThat(subscriptionIdCaptor.getValue()).isEqualTo(subscription.getSubscriptionId());
+    assertThat(responseCaptor.getValue()).usingRecursiveComparison().isEqualTo(expectedNewBlock);
   }
 
   @Test
   public void shouldReturnTxHashesWhenIncludeTransactionsFalse() {
     final NewBlockHeadersSubscription subscription = createSubscription(false);
     mockSubscriptionManagerNotifyMethod(subscription);
-    final List<Hash> txHashList = transactionsWithHashOnly();
-    final BlockResult expectedNewBlock = expectedBlockWithTransactions(txHashList);
+    final List<Transaction> transactions = transactions();
 
-    simulateAddingBlockOnCanonicalChain();
+    final Block testBlock =
+        appendBlockWithParent(blockchain, blockchain.getChainHeadBlock(), transactions);
+    simulateAddingBlockOnCanonicalChain(testBlock);
+    final BlockResult expectedNewBlock =
+        blockResultFactory.transactionHash(
+            blockchainQueriesSpy.blockByHashWithTxHashes(testBlock.getHash()).orElse(null));
 
     verify(subscriptionManager)
         .sendMessage(subscriptionIdCaptor.capture(), responseCaptor.capture());
     assertThat(subscriptionIdCaptor.getValue()).isEqualTo(subscription.getSubscriptionId());
     final Object actualBlock = responseCaptor.getValue();
     assertThat(actualBlock).isInstanceOf(BlockResult.class);
-    assertThat(((BlockResult) actualBlock).getTransactions()).hasSize(txHashList.size());
-    assertThat(actualBlock).isEqualToComparingFieldByFieldRecursively(expectedNewBlock);
+    assertThat(((BlockResult) actualBlock).getTransactions()).hasSize(transactions.size());
+    assertThat(actualBlock).usingRecursiveComparison().isEqualTo(expectedNewBlock);
 
-    verify(blockchainQueries, times(1)).blockByHashWithTxHashes(any());
-    verify(blockchainQueries, times(0)).blockByHash(any());
+    verify(blockchainQueriesSpy, times(2)).blockByHashWithTxHashes(any());
+    verify(blockchainQueriesSpy, times(0)).blockByHash(any());
   }
 
   @Test
   public void shouldReturnCompleteTxWhenParameterTrue() {
     final NewBlockHeadersSubscription subscription = createSubscription(true);
     mockSubscriptionManagerNotifyMethod(subscription);
-    final List<TransactionWithMetadata> txHashList = transactionsWithMetadata();
-    final BlockWithMetadata<TransactionWithMetadata, Hash> testBlockWithMetadata =
-        new BlockWithMetadata<>(
-            blockHeader, txHashList, Collections.emptyList(), blockHeader.getDifficulty(), 0);
+    final List<TransactionWithMetadata> transactionsWithMetadata = transactionsWithMetadata();
+
+    List<Transaction> transactions =
```

Comment:
we need to add the final keyword whenever possible

Reply:
Done.